feat(formGeneration) add info on constraint and error management (#694)
bonitasoft/bonita-doc: md/variables.md

* _formOutput_. It is a JavaScript expression returning an object. The object structure matches the contract requirements and it is filled with formInput by default. On Submit, values entered or modified by the user and aggregated in the formOutput object (as defined by the _Data sent on click_ property of the Submit button) are submitted to the process or task to fulfill the contract.
* _taskId_. It is the id of the current BPM task. You can use it as a BPM API parameter.
* _context_. It is an External API that provides references to all business variables and documents in the process instance.
* _submit_errors_list_. It is a JavaScript expression formatting the response payload to html when a submit fails. 
In some cases, other types of variables are created:
* To display an aggregated object, a Select widget is generated to display the _available values_ of the object. 

>   The variable (External API) bound to the widget is created. It queries the BDM: `../API/bdm/businessData/com.company.model.Customer?q=find&p=0&c=99`. By default it uses the `find` query with a default pagination (only first 100 objects are returned).

Reviewer: ```suggestion
  The variable (External API) bound to the widget is created. It queries the BDM. For example, when the object Invoice has an aggregated object Customer, the query is: `../API/bdm/businessData/com.company.model.Customer?q=find&p=0&c=99`. By default it uses the `find` query with a default pagination (only first 100 objects are returned).
```

> * When the business variable is edited in the form (as specified in the contract creation wizard), a UI Designer variable is created for each variable (External API). 

Reviewer: This one should come first. This is the most generic use-case.

> By example, if the contract has been created from a business data `invoice` in the process, a variable `invoice` is created in the form and its URL is set to `../{{context.invoice_ref.link}}`.

Reviewer: ```suggestion
For example, if the contract input has been created from a business variable `invoice` in the process, a variable `invoice` is created in the form and its URL is set to `../{{context.invoice_ref.link}}`.
```

> * When the business variable is edited in the form and is related to a parent in Lazy mode, the related variables are created. Reference URL is resolved using the _lazyRef_ filter: `{{invoice|lazyRef:'customer'}}` where `customer` is related to `invoice` with a lazy reference.

Reviewer: This last bullet point is not really clear to me. All the rest is clear. 
Maybe check with another reviewer about whether this is clear to him/her.

Author: I tried to make it clearer.
What do you think ?

> If `invoice` contains lazy relations, additional variables are generated for each lazy relation to resolve (using _lazyRef_ filter). By example, if `invoice` has a `customer` relation in lazy, an External API variable `invoice_customer` is added with URL set to `{{invoice|lazyRef:'customer'}}`.

Reviewer: ```suggestion
If `invoice` contains lazy relations, additional variables are generated for each lazy relation to resolve (using _lazyRef_ filter). 
For example, if `invoice` has a `customer` relation in lazy, an External API variable `invoice_customer` is added. Its URL is set to `{{invoice|lazyRef:'customer'}}`.
```

* _formInput_. It is a JSON object. Its structure is defined by the contract inputs and the attributes are initialized with default values. It could be used to set initial values for form fields. You can set the values in formInput either by editing the default values with constants (useful for test and debug) or with values from an object in an external source that has the same model (such as with a BDM external API). You can also set the initial values of a form from some other source without using formInput. However, you will then have to edit formOutput manually.
* _formOutput_. It is a JavaScript expression returning an object. The object structure matches the contract requirements and it is filled with formInput by default. On Submit, values entered or modified by the user and aggregated in the formOutput object (as defined by the _Data sent on click_ property of the Submit button) are submitted to the process or task to fulfill the contract.
* _taskId_. It is the id of the current BPM task. You can use it as a BPM API parameter.
* _context_. It is an External API that provides references to all business variables and documents in the process instance.

> * Query variables (External API). If data contains aggregation relation, a Select widget is generated with its associated _available values_ data. It uses the BDM API like this `../API/bdm/businessData/com.company.model.Product?q=find&p=0&c=99`. By default it uses the `find` query with a default pagination (only first 100 objects are returned).

Reviewer: ```suggestion
In some cases, other types of variables are created:
* To display an aggregated object, a Select widget is generated to display the _available values_ of the object. 
  The variable (External API) bound to the widget is created. It queries the BDM: `../API/bdm/businessData/com.company.model.Product?q=find&p=0&c=99`. By default it uses the `find` query with a default pagination (only first 100 objects are returned).
```

Reviewer: An example would help

> * Data variables (External API). Created in Edit mode only. Data value is retrieved from the _context_ like this `../{{context.myBusinessData_ref.link}}`

Reviewer: ```suggestion
* When the business variable is edited in the form (as specified in the contract creation wizard), a UI Designer variable is created for each variable (External API). The value is retrieved from the _context_: `../{{context.myBusinessData_ref.link}}`
```

Reviewer: An example would help

> * Lazy reference variables. Created in Edit mode only, when a data contains lazy references. Reference URL is resolved using the _lazyRef_ filter like this `{{invoice|lazyRef:'customer'}}` where `invoice` is a data variables with a lazy reference on the `customer` relation.

Reviewer: I didn't get what is actually created. So I'm trying the following suggestion, to amend:
```suggestion
* When the business variable is edited in the form and is related to a parent in Lazy mode, the related variables are created. Reference URL is resolved using the _lazyRef_ filter: `{{invoice|lazyRef:'customer'}}` where `customer` is related to `invoice` with a lazy reference.
```

Reviewer: Good to have an example

> * _submit_errors_list_. It is a JavaScript expression formatting the response payload to html when a submit fails.

Reviewer: Since this variable is always created, I would put this one above the variables created in some cases.

Reviewer: Examples would help
